bugbug/repository.py

```diff
@@ -1543,6 +1550,69 @@ def trigger_pull() -> None:
     trigger_pull()
 
 
+def get_diff(repo_path, original_hash, fix_hash) -> bytes:
+    client = hglib.open(repo_path)
+
+    current_rev = client.identify(id=True)
+
+    try:
+        client.rawcommand([b"shelve"])
+    except Exception:
+        pass
+
+    parents = client.parents(rev=fix_hash)
+    parent_of_fix = parents[0][1]
+    client.update(rev=parent_of_fix, clean=True)
+
+    graft_result = graft(
+        client, revs=[original_hash], no_commit=True, force=True, tool=":merge"
+    )
+
+    if not graft_result:
+        return b""
+
+    final_diff = client.diff(
+        revs=[fix_hash], ignoreallspace=True, ignorespacechange=True, reverse=True
+    )
+
+    client.update(rev=current_rev, clean=True)
+
+    return final_diff
+
+
+def graft(client, revs, no_commit=False, force=False, tool=":merge") -> bool:
+    """Graft changesets specified by revs into the current repository state.
+
+    Args:
+        client: The hglib client.
+        revs: A list of the hashes of the commits to be applied to the current repository state.
+        no_commit: If True, does not commit and just applies changes in working directory.
+        force: If True, forces the grafts even if the revs are ancestors of the current repository state.
+        tool: A string representing a merge tool (see `hg help merge-tools`).
+
+    Returns:
+        Boolean of graft operation result (True for success, False for failure).
+    """
+    args = hglib.util.cmdbuilder(
+        str.encode("graft"), r=revs, no_commit=no_commit, f=force, tool=tool
+    )
+
+    eh = hglib.util.reterrorhandler(args)
+
+    try:
+        client.rawcommand(args, eh=eh, prompt=auto_resolve_conflict_prompt)
+    except Exception:
+        return False
```

Comment:
Similar to https://github.com/mozilla/bugbug/pull/4159#discussion_r1624442585

Reply:
I'm not sure if logging it as `warning` would be better than `info`. It depends on the expectations. Why could it fail, and is it normal?

Reply:
When I tested the script, I never encountered any error grafting. However, to prevent the script from stopping due to a single grafting failure, I added the `hglib.error.CommandError` exception handler to continue to the next datapoint. Due to it not being a common occurrence, I think logging it as a `warning` is more appropriate.

Reply:
I would avoid using try-exception if I do not know what type of errors I could get and why. It is better to fail loudly, which allows you to fix the problem. On the other side, failing silently will make the problems harder to notice.

Reply:
Makes sense, I've removed the try-except block as we typically should not encounter any error while grafting. Done in 0d83fa73a8732189cab7cad7bdbaf33df0b101e7!